File: python/examples/music_dsl.py
"""Example Music DSL using Grammar School."""

from grammar_school import Grammar, method


class MusicGrammar(Grammar):
    """A simple music DSL for creating tracks and clips."""

    def __init__(self):
        super().__init__()
        self.tracks = []
        self.current_track = None

    @method
    def track(self, name, color=None):
        """Create a new track."""
        track = {"name": name, "color": color, "clips": [], "muted": False}
        self.tracks.append(track)
        self.current_track = track
        print(f"Created track: {name}" + (f" (color: {color})" if color else ""))

    @method
    def add_clip(self, start, length):
        """Add a clip to the current track."""
        if self.current_track is None:
            print("Error: No track selected. Create a track first.")
            return
        clip = {"start": start, "length": length}
        self.current_track["clips"].append(clip)
        print(f"Added clip: start={start}, length={length} to track '{self.current_track['name']}'")

    @method
    def mute(self):
        """Mute the current track."""
        if self.current_track is None:
            print("Error: No track selected. Create a track first.")
            return
        self.current_track["muted"] = True
        print(f"Muted track: {self.current_track['name']}")


def main():
    """Example usage of the Music DSL."""
    grammar = MusicGrammar()

    code = 'track(name="Drums").add_clip(start=0, length=8)'
    print(f"Code: {code}")
    print("\nExecuting:")
    grammar.execute(code)

    print("\n" + "=" * 50)
    code2 = 'track(name="FX", color="blue").mute()'
    print(f"Code: {code2}")
    print("\nExecuting:")
    grammar.execute(code2)

    print("\n" + "=" * 50)
    print("Final state:")
    print(f"Tracks: {grammar.tracks}")


if __name__ == "__main__":
    main()

File: go/examples/music_dsl.go
package main

import (
	"context"
	"fmt"
	"grammar-school/gs"
)

// MusicDSL is a simple music DSL for creating tracks and clips.
type MusicDSL struct{}

// Track creates a new track.
func (d *MusicDSL) Track(args gs.Args, ctx *gs.Context) ([]gs.Action, *gs.Context, error) {
	name := args["name"].Str
	color := ""
	if c, ok := args["color"]; ok {
		color = c.Str
	}

	action := gs.Action{
		Kind: "create_track",
		Payload: map[string]interface{}{
			"name":  name,
			"color": color,
		},
	}

	return []gs.Action{action}, ctx, nil
}

// AddClip adds a clip to the current track.
func (d *MusicDSL) AddClip(args gs.Args, ctx *gs.Context) ([]gs.Action, *gs.Context, error) {
	start := args["start"].Num
	length := args["length"].Num

	action := gs.Action{
		Kind: "add_clip",
		Payload: map[string]interface{}{
			"start":  start,
			"length": length,
		},
	}

	return []gs.Action{action}, ctx, nil
}

// Mute mutes the current track.
func (d *MusicDSL) Mute(args gs.Args, ctx *gs.Context) ([]gs.Action, *gs.Context, error) {
	action := gs.Action{
		Kind:    "mute_track",
		Payload: map[string]interface{}{},
	}

	return []gs.Action{action}, ctx, nil
}

// MusicRuntime is a simple runtime that prints actions.
type MusicRuntime struct{}

func (r *MusicRuntime) ExecuteAction(ctx context.Context, a gs.Action) error {
	fmt.Printf("Executing: %s with payload: %v\n", a.Kind, a.Payload)
	return nil
}

// SimpleParser is a placeholder parser (would need actual implementation).
type SimpleParser struct{}

func (p *SimpleParser) Parse(input string) (*gs.CallChain, error) {
	// This is a placeholder - in a real implementation, you'd use
	// a parser library like participle or pigeon here.
	// For now, we'll return an error indicating a parser is needed.
	return nil, fmt.Errorf("parser not implemented - use a real parser backend")
}

func main() {
	dsl := &MusicDSL{}
	parser := &SimpleParser{}
	engine, err := gs.NewEngine("", dsl, parser)
	if err != nil {
		fmt.Printf("Error creating engine: %v\n", err)
		return
	}

	runtime := &MusicRuntime{}

	// Note: This won't work until a real parser is implemented
	code := `track(name="Drums").add_clip(start=0, length=8)`
	fmt.Printf("Code: %s\n", code)
	fmt.Println("\nNote: Parser implementation needed to run this example")
	fmt.Println("See parser_backend.go for the Parser interface")

	// Once parser is implemented, uncomment:
	// plan, err := engine.Compile(code)
	// if err != nil {
	// 	fmt.Printf("Error compiling: %v\n", err)
	// 	return
	// }
	// engine.Execute(context.Background(), runtime, plan)
}
